internal/controllers/machinedeployment/mdutil/util.go

```diff
@@ -469,58 +475,71 @@ func FindNewMachineSet(deployment *clusterv1.MachineDeployment, msList []*cluste
 	// We deterministically choose the oldest new MachineSet with matching template hash.
 	sort.Sort(MachineSetsByDecreasingReplicas(msList))
 
-	var matchingMachineSets []*clusterv1.MachineSet
+	var newMSCandidates []*clusterv1.MachineSet
+	oldMSs = make([]*clusterv1.MachineSet, 0)
+	oldMSNotUpToDateResults = make(map[string]NotUpToDateResult)
 	var diffs []string
 	for _, ms := range msList {
-		upToDate, logMessages, _ := MachineTemplateUpToDate(&ms.Spec.Template, &deployment.Spec.Template)
+		upToDate, notUpToDateResult := MachineTemplateUpToDate(&ms.Spec.Template, &deployment.Spec.Template)
 		if upToDate {
-			matchingMachineSets = append(matchingMachineSets, ms)
+			newMSCandidates = append(newMSCandidates, ms)
 		} else {
-			diffs = append(diffs, fmt.Sprintf("MachineSet %s: diff: %s", ms.Name, strings.Join(logMessages, ", ")))
+			oldMSs = append(oldMSs, ms)
+			oldMSNotUpToDateResults[ms.Name] = *notUpToDateResult
+			diffs = append(diffs, fmt.Sprintf("MachineSet %s: diff: %s", ms.Name, strings.Join(notUpToDateResult.LogMessages, ", ")))
 		}
 	}
 
-	if len(matchingMachineSets) == 0 {
-		return nil, fmt.Sprintf("couldn't find MachineSet matching MachineDeployment spec template: %s", strings.Join(diffs, "; ")), nil
+	if len(newMSCandidates) == 0 {
+		return nil, oldMSs, oldMSNotUpToDateResults, fmt.Sprintf("couldn't find MachineSet matching MachineDeployment spec template: %s", strings.Join(diffs, "; "))
 	}
 
 	// If RolloutAfter is not set, pick the first matching MachineSet.
 	if deployment.Spec.Rollout.After.IsZero() {
-		return matchingMachineSets[0], "", nil
+		for _, ms := range newMSCandidates[1:] {
+			oldMSs = append(oldMSs, ms)
+			oldMSNotUpToDateResults[ms.Name] = NotUpToDateResult{
+				// No need to set log or condition message for discarded candidates, it is not used anywhere.
```

Comment:
Just to double check. multiple MS with matching spec is something that should basically never happen, right?

(+/- cases we talked about when we would change the diff logic in a weird way, but I'm asking for the case where the diff logic stays the same)

If I'm correct, maybe let's add an additional sentence to the comment here and similar below.


Additional question, the regular rollout behavior for these MS also stays the same as before this PR, correct? (this change here just makes sure that we won't consider this MS for in-place update in later PRs)